pyproject.toml

```diff
@@ -3,4 +3,177 @@ requires = []
+build_py = "setuptools.command.build_py:build_py"
+develop = "setuptools.command.develop:develop"
+dist_info = "setuptools.command.dist_info:dist_info"
+easy_install = "setuptools.command.easy_install:easy_install"
+editable_wheel = "setuptools.command.editable_wheel:editable_wheel"
+egg_info = "setuptools.command.egg_info:egg_info"
+install = "setuptools.command.install:install"
+install_egg_info = "setuptools.command.install_egg_info:install_egg_info"
+install_lib = "setuptools.command.install_lib:install_lib"
+install_scripts = "setuptools.command.install_scripts:install_scripts"
+rotate = "setuptools.command.rotate:rotate"
+saveopts = "setuptools.command.saveopts:saveopts"
+sdist = "setuptools.command.sdist:sdist"
+setopt = "setuptools.command.setopt:setopt"
+test = "setuptools.command.test:test"
+upload_docs = "setuptools.command.upload_docs:upload_docs"
+
+[project.entry-points."setuptools.finalize_distribution_options"]
+parent_finalize = "setuptools.dist:_Distribution.finalize_options"
+keywords = "setuptools.dist:Distribution._finalize_setup_keywords"
+
+[project.entry-points."distutils.setup_keywords"]
+eager_resources = "setuptools.dist:assert_string_list"
+namespace_packages = "setuptools.dist:check_nsp"
+extras_require = "setuptools.dist:check_extras"
+install_requires = "setuptools.dist:check_requirements"
+tests_require = "setuptools.dist:check_requirements"
+setup_requires = "setuptools.dist:check_requirements"
+python_requires = "setuptools.dist:check_specifier"
+entry_points = "setuptools.dist:check_entry_points"
+test_suite = "setuptools.dist:check_test_suite"
+zip_safe = "setuptools.dist:assert_bool"
+package_data = "setuptools.dist:check_package_data"
+exclude_package_data = "setuptools.dist:check_package_data"
+include_package_data = "setuptools.dist:assert_bool"
+packages = "setuptools.dist:check_packages"
+dependency_links = "setuptools.dist:assert_string_list"
+test_loader = "setuptools.dist:check_importable"
+test_runner = "setuptools.dist:check_importable"
+use_2to3 = "setuptools.dist:invalid_unless_false"
+
+[project.entry-points."egg_info.writers"]
+PKG-INFO = "setuptools.command.egg_info:write_pkg_info"
+"requires.txt" = "setuptools.command.egg_info:write_requirements"
+"entry_points.txt" = "setuptools.command.egg_info:write_entries"
+"eager_resources.txt" = "setuptools.command.egg_info:overwrite_arg"
+"namespace_packages.txt" = "setuptools.command.egg_info:overwrite_arg"
+"top_level.txt" = "setuptools.command.egg_info:write_toplevel_names"
+"dependency_links.txt" = "setuptools.command.egg_info:overwrite_arg"
+
+[tool.setuptools]
+# disabled as it causes tests to be included #2505
+# include_package_data = true
+include-package-data = false
+
+[tool.setuptools.packages.find]
+exclude = [
+	"*.tests",
+	"*.tests.*",
+	"tools*",
+	"debian*",
+	"launcher*",
+	"newsfragments*",
+	"docs",
+	"docs.*",
+]
+namespaces = true
+
+[tool.distutils.egg_info]
+tag-build = ".post"
+tag-date = 1
```

Comment:
This will probably be a breach on the [PEP 621 requirements for enforcing dynamic values](https://peps.python.org/pep-0621/#dynamic) of metadata.

Reply:
Yeah, that's unfortunate. We can move it back to setup.cfg if needed (though I'm not sure that helps with compliance). I'd like to replace this generation with a setuptools_scm version generation, after making vendoring/dependencies more viable.

Reply:
can `bumpversion` write the version to a `txt` file?

I think that in this case the correct would be to use

```toml
[project]
...
dynamic = [`version`]

...

[tool.setuptools.dynamic]
version = {file: ['somefile.txt']}
```

This way, I believe we can use `egg_info` because `version` is already marked as dynamic anyway...

Reply:
I suppose something like the following is also an alternative (although very inconvenient and that replicates functionality from `setuptools-scm`)

```diff
diff --git i/.bumpversion.cfg w/.bumpversion.cfg
index 557ae0ce3..301b32db9 100644
--- i/.bumpversion.cfg
+++ w/.bumpversion.cfg
@@ -2,5 +2,3 @@
 current_version = 69.5.1
 commit = True
 tag = True
-
-[bumpversion:file:setup.cfg]
diff --git i/pyproject.toml w/pyproject.toml
index ab8b02bff..eddb48898 100644
--- i/pyproject.toml
+++ w/pyproject.toml
@@ -5,7 +5,6 @@ backend-path = ["."]

 [project]
 name = "setuptools"
-version = "69.5.1"
 authors = [
        { name = "Python Packaging Authority", email = "distutils-sig@python.org" },
 ]
@@ -25,6 +24,7 @@ classifiers = [
 keywords = ["CPAN PyPI distutils eggs package management"]
 requires-python = ">=3.8"
 dependencies = []
+dynamic = ["version"]

 [project.urls]
 Homepage = "https://github.com/pypa/setuptools"
@@ -168,12 +168,3 @@ exclude = [
        "docs.*",
 ]
 namespaces = true
-
-[tool.distutils.egg_info]
-tag-build = ".post"
-tag-date = 1
-
-[tool.distutils.sdist]
-formats = "zip"
-
-[tool.setuptools_scm]
diff --git i/setup.py w/setup.py
index 542edaea6..f89dcbe37 100755
--- i/setup.py
+++ w/setup.py
@@ -1,8 +1,12 @@
 #!/usr/bin/env python

+import email
 import os
+import re
 import sys
 import textwrap
+import time
+import subprocess

 import setuptools
 from setuptools.command.install import install
@@ -81,7 +85,42 @@ class install_with_pth(install):
             self.install_lib = self.install_libbase


+def get_version() -> str:
+    if os.path.exists("PKG-INFO"):
+        with open("PKG-INFO") as fp:
+            return email.message_from_file(fp)["Version"]
+    return get_version_git() or get_version_fallback()
+
+
+def get_version_git() -> "str | None":
+    r"""
+    >>> re.match(r"^\d+\.\d+\.\d+.*", get_version_git()) != None
+    True
+    """
+    try:
+        cmd = ["git", "describe", "--tags", "--match", "v*.*.*"]
+        version = subprocess.check_output(cmd, text=True, encoding="utf-8").lstrip("v")
+        return version.replace("-", ".post", 1).replace("-", "+", 1)
+    except subprocess.CalledProcessError as ex:
+        print(f"Cannot find git version: {ex}\n{ex.stdout}\n{ex.stderr}")
+        return None
+
+
+def get_version_fallback() -> str:
+    r"""
+    >>> re.match(r"^\d+\.\d+\.\d+.*", get_version_fallback()) != None
+    True
+    """
+    with open(".bumpversion.cfg", encoding="utf-8") as f:
+        match = re.search(r"(?M)current_version = (?P<version>[.0-9]+)", f.read())
+        if not match:
+            print("Cannot find version in .bumpversion.cfg")
+            return "0.0.0+error"
+        return f"{match.group('version')}.post{time.strftime('%Y%m%d')}"
+
+
 setup_params = dict(
+    version=get_version(),
     cmdclass={'install': install_with_pth},
     package_data=package_data,
 )
```

Reply:
The nice thing about using setuptools_scm is that it produces a viable pre-release version that's greater than the last release and increases with each commit:

```
 setuptools feature/pep-621 @ pip-run setuptools_scm -- -m setuptools_scm
69.5.2.dev167+g1b8981295
```

Since the last release was 69.5.1, it creates 69.5.2, adds the dev plus a distance since that release (167 commits) and then tags it with the git hash (to distinguish from another commit 167 commits from 69.5.1). If there were uncommitted changes, there'd be an additional local version identifier representing the hash of those changes. This approach provides some fidelity as the code evolves between versions. There's no way that bumpversion can do that, because bumpversion is only run once per release (and that's once too many in my opinion).

And while it's possible that we could essentially vendor some subset of setuptools_scm in setuptools' own setup.py, that seems like the wrong path. Why wouldn't we want to re-use the official sanctioned plugin and build on the lessons and functionality it's developed over the years (if viable)?

Reply:
By the way, I'm assuming this change isn't a blocker for now. Are you okay with me merging this, even though it violates the dynamic specification?

Reply:
I understand... I use `setuptools-scm` for the same reason in all my projects.

My concern with `dynamic` is that it could affect the dependency resolution procedure (in theory it is possible for the installer to trust `pyproject.toml` without reading `METADATA`). But then, there is also the version inference by file name throwed at the mix...

If we are happy with the keeping the non-ideal `bumpversion` solution https://github.com/pypa/setuptools/pull/4310#discussion_r1604500026 might be a way of conciliate it with, without violating the `dynamic` assumption.

---

Should we hold the release until https://github.com/pypa/setuptools/pull/4360[^1]? Or just go with it now?

The only reason why I did not cut a release yet, was because of I was trying to avoid 2 major releases one after the other .

[^1]: Reasoning in https://github.com/pypa/setuptools/discussions/4345#discussioncomment-9447827

Reply:
Oh, gosh. I don't think I'd realized until now that the bumpversion technique was broken (it still mentions setup.cfg). Yeah, we'll need to do something just to keep the release working. Is there any reason we need to re-engineer the version bump and not simply follow the same pattern as before just with `pyproject.toml` instead of `setup.cfg`? In other words, why not use 491f2ca27 and not change anything else? I did test after that commit and the `tox -e finalize` does run successfully. Can we defer any changes to version management to a separate issue/PR?

Reply:
> Yes, definitely. I think we can merge this right now. But I think it would be nice if we solve the PEP 621 conflict before releasing.

I'd like to avoid merging something that blocks the release.

The way this logic was designed is for these flags to be the default when developing, but when cutting a release, the values are overridden to use just the version statically.

I just realized, though, that the release process depends on writing to setup.cfg to override the defaults, and unless setup.cfg takes precedence over pyproject.toml, it's actually not going to work, meaning a tagged release will have the tags on the version.

Since I want to focus on getting the majority of the config moved to pyproject.toml, I'm going to instead move these settings back to setup.cfg so we can get most of the config moved but remain stable with respect to these tags.

Reply:
Done in e775f9906.

Reply:
> The way this logic was designed is for these flags to be the default when developing, but when cutting a release, the [values are overridden](https://github.com/pypa/setuptools/blob/5cbf12a9b63fd37985a4525617b46576b8ac3a7b/tox.ini#L104-L105) to use just the version statically.

Oh, I see. It should be fine then!

Thank you very much for having a look on this Jason.